### PR TITLE
MM-61130: Use a channelMember map at web_hub level

### PR DESCRIPTION
**Pull request by agnivade**
Tests at very high scale indicates that the iteration
of all connections during websocket broadcast starts
to become a bottleneck.

To optimize this, we move the channelMember cache from
inside web_conn.go to the hubConnectionIndex.

This involves adding a new map keyed by the channelID
and containing all webConns where the user is a member
of that channel. Subsequently, a new method needed to
be added to invalidate the cache which previously
used to happen in web_conn.

And as a last step, we remove the cache from web_conn
to reduce SQL queries to the DB.

https://mattermost.atlassian.net/browse/MM-61130

Attached are the cpu profiles: 
[cpu_profiles.zip](https://github.com/user-attachments/files/17427714/cpu_profiles.zip)

cpu_150 is old one
cpu_180 is with the changes

Grafana:
OLD:
Threshold: https://snapshots.raintank.io/dashboard/snapshot/mZajUP5Ereb4H55Wmp792BOsMdTMUjQC?orgId=0
V2: https://snapshots.raintank.io/dashboard/snapshot/YiNXaWhhSJGYHjRBbBII1tmlCbWn9fEL?orgId=0

NEW:
Threshold: https://snapshots.raintank.io/dashboard/snapshot/DfqeF49pVt1raQ4jTWzxMmgGXWQjgUbL
load test: (attached)
sc3
sc2
sc1
sc6
sc4

```release-note
NONE
```

### CODE REVIEW

File: server/channels/app/platform/web_conn.go
Comment (cpoile) [inline server/channels/app/platform/web_conn.go:326]: A little bit unrelated, but do we _really_ need this to be an atomic? It's only set on connect and reconnect, so maybe we can save some sync costs? (Do we have tests that would detect an error on reconnect if we did remove this?) Undestandable if we don't want to introduce the chance of a race here, but I'm just thinking of ways to squeeze more perf out of this.

**Review by cpoile — APPROVED** [review]
This is some nice code, thanks Agniva!
The only observation I would have is that I'm worried the tests are really stressing the combinations of maps we now have. (Unless that's somewhere else?) It would be nice to see, say, a function that took 100 users over 10 channels and added and removed them hundreds of times randomly, and then asserted that the final combination of users/channels is correctly stored in the various maps. Or maybe that's too idealistic. Anyway, not a req. Nice work!

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [inline server/channels/app/platform/web_hub.go:675]: I am still not happy with this. Thinking more, if this failed inside web_conn.go, only one connection gets affected, but if this fails here, then a large portion of users will be affected. We cannot continue with a failure. Will improve this.

File: server/channels/app/platform/web_conn.go
Comment (agnivade) [reply server/channels/app/platform/web_conn.go:326]: Hmm .. I need to check this. Atomics are usually extremely fast, and this has never appeared in the profiles. But I am totally supportive of not using atomics unless we need to. I'll look into this separately.

**Comment by agnivade** [comment]
>  It would be nice to see, say, a function that took 100 users over 10 channels and added and removed them hundreds of times randomly, and then asserted that the final combination of users/channels is correctly stored in the various maps. Or maybe that's too idealistic. 

This is a good idea. The problem is app/platform doesn't have (or rather, difficult to add) the necessary helper functions (due to refactors) to do such high level things. I even needed to copy-paste the store method to add a channel member, because suiteIface didn't have that.

I will add a test from api layer to check this. Good point.

**Comment by agnivade** [comment]
Ok, this is ready to review now, sans the high level additional tests with cpoile asked for.

**Comment by streamer45** [comment]
@agnivade A meta request for this would be (if possible) to link to some dashboards (i.e., Grafana snapshot) or supporting data from scaling tests you are conducting (e.g., pprof profiles).

**Comment by agnivade** [comment]
@streamer45 - Done. Attached the results in the PR message.

**Comment by agnivade** [comment]
Added high level test @cpoile. PTAL.

**Review by cpoile — APPROVED** [review]
I'm assuming there's no chance that adding and removing multiple users would cause inconsistency, right? I think this is good then, thanks Agniva!

**Comment by agnivade** [comment]
It's really the same piece of code that gets executed. Don't think that increases code coverage anyway. Important thing is that we covered the add and remove paths.

File: server/channels/app/platform/web_hub.go
Comment (agarciamontoro) [inline server/channels/app/platform/web_hub.go:746]: We're `panic`-ing here as well, I think that's another argument to `panic` inside the function.

File: server/channels/app/platform/web_hub.go
Comment (agarciamontoro) [inline server/channels/app/platform/web_hub.go:768]: Are we concerned about this millisecond here? Also, should we log the first error in the query in case it does affect performance?

File: server/channels/app/platform/web_hub.go
Comment (agarciamontoro) [inline server/channels/app/platform/web_hub.go:804]: Why do we have a special case for `len(i.byChannelID[channelID]) <= 1`? If we need to clone it, why not clone it there as well?

**Review by agarciamontoro — APPROVED** [review]
Just a couple of questions and nits below. This is a great improvement!

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:746]: Yep, good point.

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:768]: Not particularly concerned. It's a hit we need to take. Will log the first error. Thanks.

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:804]: Because the slice length is not re-adjusted there. Since it has just one element. Refer to the PR: https://github.com/mattermost/mattermost/pull/25785

File: server/channels/app/platform/web_hub.go
Comment (streamer45) [inline server/channels/app/platform/web_hub.go:758]: I'd prefer if we prevented the specific user from connecting rather than crashing the hub along with all connections (from multiple users) in them.

File: server/channels/app/platform/web_hub.go
Comment (streamer45) [reply server/channels/app/platform/web_hub.go:768]: Not sure I fully understand the use of `Sleep` here. The caller of `getChannelMembersForUser` is blocked regardless, so what exactly are we unblocking by sleeping?

File: server/channels/app/platform/web_hub.go
Comment (streamer45) [inline server/channels/app/platform/web_hub.go:728]: This code seems practically identical to the above, except for the filtering condition. Wondering if it'd be valuable to refactor into a single util.

File: server/channels/app/platform/web_hub.go
Comment (streamer45) [inline server/channels/app/platform/web_hub.go:752]: Maybe I'd move this on top of the file and give it a slightly more verbose name to clearly indicate the context of its purpose.

File: server/channels/app/platform/web_hub.go
Comment (streamer45) [inline server/channels/app/platform/web_hub.go:797]: Wondering if the fast path here is adding much value. It feels like additional code to cover an unrealistic case that would be fast regardless. Thoughts?

File: server/channels/app/platform/web_hub.go
Comment (streamer45) [inline server/channels/app/platform/web_hub.go:505]: To confirm, this is the only place where we need to invalidate, and it would trigger if a user joined a new channel or left an existing one (and a few more cases).

File: server/channels/app/platform/web_hub.go
Comment (streamer45) [inline server/channels/app/platform/web_hub.go:677]: Mostly a curiosity of mine: do we have heap profiles around to understand how heavy this map can get?

**Review by streamer45 — COMMENTED** [review]
Great improvement. Left a few comments and thoughts.

I'd also recommend security and qa reviews since we are touching the scope of ws events, but will defer to you on that.

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:768]: By letting the DB some time to recover from whatever it is failing from.

Comment (streamer45) [reply server/channels/app/platform/web_hub.go:768]: I see. I find the comment a bit unclear if the purpose of the sleep is to let the DB "breath". Plus I am not sure one millisecond would help that much since we are talking about a couple of retries so if it failed a ms before, I say there's a good chance it could fail again. And now I wonder, how does the retry store layer fit into this? In that case we pause for 100ms which sounds more reasonable.

Comment (agnivade) [reply server/channels/app/platform/web_hub.go:768]: The retry store layer is purely for retriable queries. e.g: ones which got cancelled due to a deadlock. Those are safe to be auto-retried. This is more about if the DB failed to reply due to load or something. So what you are suggesting is to remove the retry completely? I'd be fine with that.

Comment (streamer45) [reply server/channels/app/platform/web_hub.go:768]: Yes, it could be simpler if we don't retry at all and fail fast. Otherwise, we'd have to bump the sleep to a more meaningful value, I feel, and block for longer as a result.

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:505]: Correct. We are just doing it at the same level as the `webConn.InvalidateCache`, because we are basically hoisting up the membership check from webConn level to webHub level.

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:571]: Done

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:677]: Very minimal. 4.25MB. From the latest test.
[heap_200K.zip](https://github.com/user-attachments/files/17537000/heap_200K.zip)

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:728]: Let's do it if this gets repeated one more time. Following the rule of 3.

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:768]: Done.

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:797]: Makes sense. Removing.

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:758]: That will be difficult at the hub level because this data is used to control broadcast data for all users in the hub. The per-user control was possible when this was at the web_conn level.

**Comment by agnivade** [comment]
@streamer45 - PTAL. This has become more complicated now. I ran into situations where a small mistake would lead to a hub deadlock in trying to force close the connection.

@agarciamontoro - requesting another review as well because the code has changed a lot after Claudio's suggestion to force-close the user connection instead of panicking.

I am still not 100% confident because this code is one of the most complicated areas in MM. But let me know how it looks to you guys. We definitely need a thorough QA and security review on this.

**Comment by agnivade** [comment]
Hmm, ok the tests are stuck. There's probably a deadlock somewhere.

**Comment by agnivade** [comment]
I cancelled the tests after 50 minutes. Pretty sure there's a deadlock somewhere. 

 @streamer45 @agarciamontoro - What are your thoughts on dropping this PR? Do you think the gain is worth the risk?

File: server/channels/app/platform/web_hub.go
Comment (streamer45) [inline server/channels/app/platform/web_hub.go:277]: It looks safe as is, but how about making this buffered to avoid blocking the hub even for a moment?

File: server/channels/app/platform/web_hub.go
Comment (streamer45) [inline server/channels/app/platform/web_hub.go:527]: This could be problematic as I think closing would call `(h *Hub) Unregister` before returning.

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:277]: This feels like a micro-optimization. I want to lean on the side of correctness for now.

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go]: You are right. Following the existing method of :

```
close(webConn.send)
connIndex.Remove(webConn)
```

**Review by streamer45 — COMMENTED** [review]
Thanks. I understand that working with mocks has been painful on this one but I also think it would be valuable to have test cases for `GetAllChannelMembersForUser` failing. Anything we can do about that?

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:526]: Yes, we copy the slice, so it should be fine.

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:528]: Fair enough :P

**Comment by agnivade** [comment]
>  I understand that working with mocks has been painful on this one but I also think it would be valuable to have test cases for GetAllChannelMembersForUser failing. Anything we can do about that?

This would be really challenging. Assigning the mock store to the hub causes all sorts of race conditions which are hard to eradicate. I personally tested this by randomly failing the store method once every 4 calls and verified that the websocket reconnects fine. Did this both for connect and invalidate paths.

File: server/channels/app/platform/web_hub.go
Comment (agnivade) [reply server/channels/app/platform/web_hub.go:528]: Hmm .. I only see this 2 times. Where are you seeing the 3rd time?

Comment (agnivade) [reply server/channels/app/platform/web_hub.go:528]: Ah nvm :facepalm:

**Comment by agnivade** [comment]
Ok, added a test, and found a bug! :scream:

File: server/channels/api4/post_test.go
Comment (agnivade) [reply server/channels/api4/post_test.go:3081]: Indeed. :D

**Comment by agnivade** [comment]
Gentle ping @agarciamontoro.

**Review by agarciamontoro — APPROVED** [review]
Changes look good to me, I'm happy we're no longer `panic`ing :)

**Comment by DHaussermann** [comment]
@streamer45 and @agnivade I'm having a setup issue with this. As per test steps provided in the Jira, I need my local HA setup working but it does not want to start the way it has in the past for other PRs
image

I'm confused because I think these are different images that run in AMD64 so what is unique about the inbucket image?

0/5 Is there a quick way I can remove inbucket from the setup to see if it resolves the problem or is it required for HA?

**Comment by DHaussermann** [comment]
Regarding the setup issue ☝️ the local HA command seems unreliable. I got this working simply by trying the same thing repeatedly. I don't know why it sometimes cause a problem that inbucket doesn't work and sometimes does. 
Also the Redis container sometimes does not start for no apparent reason and I must start it manually and then restart the server.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- I tested this locally on HA having 3 - 4 unique users and sessions and monitoring WS events
- Completed all actions listed in the jira task for add/remove users to teams and channels
- Using events such as the `typing`, I ensured events were scoped properly based on channel membership and team membership
- Ensured no issues after a connection drop and restore
- Repeated testing with Redis disabled
No issues found. 

LGTM!

**Comment by agnivade** [comment]
Thanks Dylan!

> Regarding the setup issue ☝️ the local HA command seems unreliable. 

Could you create a post in hub in PDE private discussion and see what other people say? I don't use the Makefile commands, but if this is unreliable, then we definitely need to make this correct.

**Pull request merged**
